rclcpp_action/include/rclcpp_action/client.hpp

```diff
@@ -342,6 +342,20 @@ class Client : public ClientBase
   std::shared_future<typename GoalHandle::SharedPtr>
   async_send_goal(const Goal & goal, const SendGoalOptions & options = SendGoalOptions())
   {
+    // To prevent the list from growing out of control, forget about any goals
```

Comment:
I would be happier with this if:
- It were a separate, well named function that could be called by the object using the client to clear itself if it wanted to
- This cleaning up process, if happening in the send goal command, happened at the end of it so it didn't add any additional latency to sending the command, even if the client thread had a little latency for returning. That lock makes me nervous. 
- add a default to `true` parameter in async_send_goal whether or not it wants this method to do the housekeeping for you. Maybe someone wants to do it themselves.

Reply:
> It were a separate, well named function that could be called by the object using the client to clear itself if it wanted to

Good idea. But in order to try and get this fix into Foxy, I think I'll just add a TODO to expose a function. This way we avoid adding/modifying the existing API.

> add a default to `true` parameter in async_send_goal whether or not it wants this method to do the housekeeping for you.

Ditto. I'd like to postpone changes to API until post-Foxy.

> This cleaning up process, if happening in the send goal command, happened at the end of it so it didn't add any additional latency to sending the command, even if the client thread had a little latency for returning. That lock makes me nervous.

I can do this.